tensorflow/core/kernels/non_max_suppression_op.cu.cc

```diff
@@ -0,0 +1,433 @@
+            flipped.y2 = box.y1;
+          }
+        }
+        shared_i_boxes[threadIdx.x] = flipped;
+        shared_i_areas[threadIdx.x] =
+            (flipped.x2 - flipped.x1 + 1.0f) * (flipped.y2 - flipped.y1 + 1.0f);
+      }
+    }
+    __syncthreads();
+    const int i = i_block_offset + threadIdx.x;
+    for (int j_thread_offset =
+             NMS_BOXES_PER_THREAD * (blockIdx.y * blockDim.y + threadIdx.y);
+         j_thread_offset < nboxes;
+         j_thread_offset += NMS_BOXES_PER_THREAD * blockDim.y * gridDim.y) {
+      // Note : We can do everything using multiplication,
+      // and use fp16 - we are comparing against a low precision
+      // threshold
+      int above_thresh = 0;
+      bool valid = false;
+      for (int ib = 0; ib < NMS_BOXES_PER_THREAD; ++ib) {
+        // This thread will compare Box i and Box j
+        const int j = j_thread_offset + ib;
+        if (i < j && i < nboxes && j < nboxes) {
+          valid = true;
+          const Box o_box = d_desc_sorted_boxes[j];
+          const Box i_box = shared_i_boxes[threadIdx.x];
+          Box j_box = o_box;
+          if (flip_boxes) {
+            if (j_box.x1 > j_box.x2) {
+              float tmp = j_box.x2;
+              j_box.x2 = j_box.x1;
+              j_box.x1 = tmp;
+            }
+            if (j_box.y1 > j_box.y2) {
+              float tmp = j_box.y2;
+              j_box.y2 = j_box.y1;
+              j_box.y1 = tmp;
+            }
+          }
+          const float j_area =
+              (j_box.x2 - j_box.x1 + 1.0f) * (j_box.y2 - j_box.y1 + 1.0f);
+          const float i_area = shared_i_areas[threadIdx.x];
+          // The following code will not be valid with empty boxes
+          if (i_area == 0.0f || j_area == 0.0f) continue;
+          const float xx1 = fmaxf(i_box.x1, j_box.x1);
+          const float yy1 = fmaxf(i_box.y1, j_box.y1);
+          const float xx2 = fminf(i_box.x2, j_box.x2);
+          const float yy2 = fminf(i_box.y2, j_box.y2);
+
+          // fdimf computes the positive difference between xx2+1 and xx1
+          const float w = fdimf(xx2 + 1.0f, xx1);
+          const float h = fdimf(yy2 + 1.0f, yy1);
+          const float intersection = w * h;
+
+          // Testing for a/b > t
+          // eq with a > b*t (b is !=0)
+          // avoiding divisions
+          const float a = intersection;
+          const float b = i_area + j_area - intersection;
+          const float bt = b * thresh;
+          // eq. to if ovr > thresh
+          if (a > bt) {
+            // we have score[j] <= score[i]
+            above_thresh |= (1U << ib);
+          }
+        }
+      }
+      if (valid) {
+        d_delete_mask[i * mask_ld + j_thread_offset / NMS_BOXES_PER_THREAD] =
+            above_thresh;
+      }
+    }
+    __syncthreads();  // making sure everyone is done reading smem
+  }
+}
+
+template <typename T, typename Index>
+__global__ void IndexSelect(const CudaLaunchConfig config, const T* original,
```

Comment:
I would probably just pass in virtual_thread_count (call it num_elements or something).